dap_job_quality/pipeline/sentence_classifier/prep_training_data.py

```diff
@@ -1,98 +1,318 @@
 
-    # Find sentences that are not in the 'sentence' column
-    new_sentences = all_sentences_from_text - existing_sentences
+def process_additional_data():
+    additional_data = get_additional_examples_underrepresented_categories()
 
-    return new_sentences
+    # We don't need this column - it contains occasional notes for just a few rows
+    additional_data = additional_data.drop(columns=["Unnamed: 13"])
 
+    # Replace "different category" in "JQ" with values from "correct_category"
+    additional_data.loc[
+        additional_data["JQ"] == "different category", "JQ"
+    ] = additional_data["correct_category"]
 
-def filter_job_ads(labelled_df):
-    job_ids = labelled_df["id"].unique()
+    additional_data.loc[additional_data["JQ"] == "1", "JQ"] = additional_data[
+        "subcategory"
+    ]
 
-    # skip the first 10 job ads - we didn't know what we were labelling at that point
-    target_ids = job_ids[10:]
+    additional_data["label"] = additional_data["JQ"].apply(
+        lambda x: 0 if x == "0" or x == "not sure" else 1
+    )
+
+    return additional_data
+
+
+def find_category_representation(additional_data):
+    additional_positives = additional_data[additional_data["label"] == 1]
+    # Drop the column "subcategory"
+    additional_positives = additional_positives.drop(columns=["subcategory"])
+
+    # Rename the column "JQ" to "subcategory"
+    additional_positives = additional_positives.rename(columns={"JQ": "subcategory"})
+
+    positive_categories = pd.concat(
+        [
+            positive_sents[["id", "subcategory"]],
+            additional_positives[["id", "subcategory"]],
+        ]
+    )
+
+    logging.info(f"Total number of positive sentences: {len(positive_categories)}")
+
+    counts = positive_categories["subcategory"].value_counts()
+
+    proportions = positive_categories["subcategory"].value_counts(normalize=True)
+
+    combined_df = pd.DataFrame(
+        {
+            "subcategory": counts.index,
+            "count": counts.values,
+            "proportion": proportions.values,
+        }
+    )
+
+    logging.info(f"Category representation: {combined_df}")
+
+    combined_df.to_csv(
+        PROJECT_DIR / "outputs/data/category_representation_in_positive_sentence.csv",
+        index=False,
+    )
+
+    return positive_categories, additional_positives
+
+
+def get_negative_example_sentences(
+    df, id, sentences_col="sentences", true_sent_col="sentence"
+):
+    subset = df[df["id"] == id].copy()
+
+    sentences = set(subset[sentences_col].iloc[0])
+    sentences_w_spans = set(subset[true_sent_col].unique())
+    neg_sentences = sentences - sentences_w_spans
+    return neg_sentences
+
+
+def create_neg_sent_df(positive_sents, false_negatives, additional_data):
+    negative_sentences = {}
+
+    for id in positive_sents["id"].unique():
+        neg_sentences = get_negative_example_sentences(positive_sents, id)
+        negative_sentences[id] = list(neg_sentences)
```

Comment:
I think using a groupby will make this a bit cleaner
```suggestion
def get_negative_example_sentences(
    subset, sentences_col="sentences", true_sent_col="sentence"
):
    sentences = set(subset[sentences_col].iloc[0])
    sentences_w_spans = set(subset[true_sent_col].unique())
    neg_sentences = sentences - sentences_w_spans
    return neg_sentences


def create_neg_sent_df(positive_sents, false_negatives, additional_data):
    negative_sentences = {}

    for id, subset in positive_sents.groupby("id"):
        neg_sentences = get_negative_example_sentences(subset)
        negative_sentences[id] = list(neg_sentences)
```